pkg/plugin/resource/resource.go

```diff
@@ -0,0 +1,200 @@
+			for _, d := range descriptions {
+				ids = append(ids, d.ID)
+			}
+			return "", fmt.Errorf("Found multiple resources named %s: %v", name, ids)
+		}
+	}
+	numFound := len(resourceIDs)
+
+	for _, name := range orderedResourceNames {
+		if _, ok := resourceIDs[name]; ok {
+			continue
+		}
+
+		resource := resources[name]
+		properties, err := executeAsTemplate(json.RawMessage(*resource.config.Properties), struct{ Resources interface{} }{resourceIDs})
+		if err != nil {
+			return "", fmt.Errorf("Failed to get properties for %s: %s", name, err)
+		}
+
+		id, err := resource.plugin.Provision(instance.Spec{
+			Properties: &properties,
+			Tags:       map[string]string{resourceGroupTag: string(config.ID), resourceNameTag: name},
+		})
+		if err != nil {
+			return "", fmt.Errorf("Failed to provision %s: %s", name, err)
+		}
+		log.Infof("Provisioned %s with ID %s", name, *id)
+		resourceIDs[name] = struct{ instance.ID }{*id}
+	}
+
+	return fmt.Sprintf("Found %d resources and created %d ", numFound, len(resourceIDs)-numFound), nil
+}
+
+func executeAsTemplate(text json.RawMessage, data interface{}) (json.RawMessage, error) {
+	tmpl, err := template.New("").Option("missingkey=error").Parse(string(text))
+	if err != nil {
+		return nil, err
+	}
+
+	var b bytes.Buffer
+	if err = tmpl.Execute(&b, data); err != nil {
+		return nil, err
+	}
+
+	return b.Bytes(), nil
+}
+
+var resourceReferenceRegexp = regexp.MustCompile(`{{\s*\.Resources\.(\w+)`)
+
+func getResourceReferences(properties json.RawMessage) []string {
+
+	var references []string
+	// TODO: Use text/template.Template.Execute instead.
+	for _, submatches := range resourceReferenceRegexp.FindAllSubmatch(properties, -1) {
+		references = append(references, string(submatches[1]))
+	}
+	return references
+}
+
+func getProvisioningOrder(resources map[string]*resource) ([]string, error) {
+	g := graph.New(graph.Directed)
+
+	nodes := map[string]graph.Node{}
+	for name := range resources {
+		nodes[name] = g.MakeNode()
+		*nodes[name].Value = name
+	}
+
+	for name, resource := range resources {
+		to := nodes[name]
+		references := getResourceReferences(json.RawMessage(*resource.config.Properties))
+		for _, reference := range references {
+			from, ok := nodes[reference]
+			if !ok {
+				return nil, fmt.Errorf("Reference to undefined resource %s in %s", reference, name)
+			}
+			if err := g.MakeEdge(from, to); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	for _, connectedComponent := range g.StronglyConnectedComponents() {
+		if len(connectedComponent) > 1 {
+			var cycle []string
+			for _, node := range connectedComponent {
+				cycle = append(cycle, (*node.Value).(string))
+			}
+			sort.Strings(cycle)
+			return nil, fmt.Errorf("Cycle detected: %v", cycle)
+		}
+	}
+
+	var sorted []string
+	for _, node := range g.TopologicalSort() {
+		sorted = append(sorted, (*node.Value).(string))
+	}
+
+	return sorted, nil
+}
+
+func (p *plugin) FreeGroup(id group.ID) error {
```

Comment:
Considering the resource plugin *delegates* resource CRD (create/read/destroy) to the instance plugin, using `instance.ID` may make sense.

On the issue of `Destroy` -- I am not so sure if destroying individual resources (the `Destroy` method with an id) makes a lot of sense now...  How about a `Destroy(spec *resource.Spec, pretend bool)` instead?   This is effectively the *inverse* of commit and we should be able to implement this by computing the *reverse* order of resource creation.    This should also work around the problem of maintaining association of Id to the instance plugin, right?

One thing we are ignoring for now is the issue that some resources cannot be destroyed if the code is running from *within* or *on* that resource -- eg. destroying a VPC in which the host that calls the `destroy` is running.   

What do you think?  Can you make this change and add the implementation to reverse the topological sort for destroying all the resources?